### PR TITLE
[MU4] Languages page

### PR DESCRIPTION
**Pull request by Eism**
Peek 2020-08-04 10-18

### CODE REVIEW

File: mu4/languages/internal/languagescontroller.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+void LanguagesController::th_install(const QString& languageCode, async::Channel<LanguageProgress> progressChannel,
+                                     std::function<void(const QString&, const Ret&)> callback)
+{
+    progressChannel.send(LanguageProgress(ANALYSING_STATUS, true));
+
+    RetVal<QString> download = downloadLanguage(languageCode, progressChannel);
+    if (!download.ret) {
+        callback(languageCode, download.ret);
```
Comment (igorkorsukov) [inline mu4/languages/internal/languagescontroller.cpp:355]: Here may be a problem.
Callback will be called in this thread, and not class thread (where set callback), implementation of callback maybe don't expect this (no mutex etc). 
I suggestion use channel for notification about finished and result.

File: mu4/appshell/qml/HomePage/AddonsContent.qml [hunk 2]
```diff
@@ -105,15 +106,16 @@ Rectangle {
...
+        LanguagesContent {
             id: languagesComp
-            color: ui.theme.backgroundColor
-            StyledTextLabel {
-                anchors.fill: parent
-                verticalAlignment: Text.AlignVCenter
-                horizontalAlignment: Text.AlignHCenter
-                text: "Languages Module"
+
+            Connections {
```
Comment (vpereverzev) [inline mu4/appshell/qml/HomePage/AddonsContent.qml:113]: But it seems that you can use it in simple property binding

`LanguagesContent {
    search: searchField.searchText
}`

File: mu4/appshell/qml/HomePage/AddonsContent.qml [hunk 3]
```diff
@@ -105,15 +106,16 @@ Rectangle {
                 }
             }
         }
-
-        Rectangle {
+        
+        LanguagesContent {
```
Comment (vpereverzev) [inline mu4/appshell/qml/HomePage/AddonsContent.qml:110]: BTW I'd prefer to call in LaguagesPage, LanguagesTab or something similar

File: mu4/languages/ilanguagesconfiguration.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    virtual QUrl languagesUpdateUrl() const = 0;
+    virtual QUrl languageFileServerUrl(const QString& languageCode) const = 0;
+
+    virtual ValCh<LanguagesHash> languages() const = 0;
+    virtual Ret setLanguages(const LanguagesHash& languages) const = 0;
+
+    virtual QString languagesSharePath() const = 0;
+    virtual QString languagesDataPath() const = 0;
+
+    virtual QStringList languageFilePaths(const QString& languageCode) const = 0;
```
Comment (vpereverzev) [inline mu4/languages/ilanguagesconfiguration.h:48]: Could you explain what is the purpose of this method?

I mean what kind of "paths" are you expecting as an output?

File: mu4/languages/ilanguagesconfiguration.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    virtual Ret setCurrentLanguageCode(const QString& languageCode) const = 0;
+
+    virtual QUrl languagesUpdateUrl() const = 0;
+    virtual QUrl languageFileServerUrl(const QString& languageCode) const = 0;
+
+    virtual ValCh<LanguagesHash> languages() const = 0;
+    virtual Ret setLanguages(const LanguagesHash& languages) const = 0;
+
+    virtual QString languagesSharePath() const = 0;
+    virtual QString languagesDataPath() const = 0;
```
Comment (vpereverzev) [inline mu4/languages/ilanguagesconfiguration.h:46]: It's not very clear what is the difference between share path and data path

Perhaps we need a comment here?

File: mu4/languages/ilanguagescontroller.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+public:
+    virtual ~ILanguagesController() = default;
+
+    virtual Ret refreshLanguages() = 0;
+
+    virtual ValCh<LanguagesHash> languages() = 0;
+    virtual RetCh<LanguageProgress> install(const QString& languageCode) = 0;
+    virtual Ret uninstall(const QString& languageCode) = 0;
+
+    virtual Ret setCurrentLanguage(const QString& languageCode) = 0;
```
Comment (vpereverzev) [inline mu4/languages/ilanguagescontroller.h:42]: well, I can see the method to "SET" current language, but I'm wondering how can I "GET" the current language? :)

File: mu4/languages/ilanguagesconfiguration.h (hunk 4 above)
Comment (Eism) [reply mu4/languages/ilanguagesconfiguration.h:48]: The language consists of several files. Here I am returning all file paths.

File: mu4/languages/ilanguagesconfiguration.h (hunk 5 above)
Comment (Eism) [reply mu4/languages/ilanguagesconfiguration.h:46]: I don't think a comment is needed here. The global configuration has two methods: sharePath and dataPath. This configuration returns these paths for languages. These are common terms.

File: mu4/languages/internal/languagescontroller.cpp (hunk 1 above)
Comment (Eism) [reply mu4/languages/internal/languagescontroller.cpp]: fixed

File: mu4/appshell/qml/HomePage/AddonsContent.qml (hunk 2 above)
Comment (Eism) [reply mu4/appshell/qml/HomePage/AddonsContent.qml]: fixed

File: mu4/appshell/qml/HomePage/AddonsContent.qml (hunk 3 above)
Comment (Eism) [reply mu4/appshell/qml/HomePage/AddonsContent.qml]: fixed

File: mu4/languages/ilanguagescontroller.h (hunk 6 above)
Comment (Eism) [reply mu4/languages/ilanguagescontroller.h:43]: fixed

**Pull request merged**